File: android/app/src/main/kotlin/com/only1cent/moteur_gr/TrekWidget.kt
// E5.19b — Widget Home Screen Android — progression trek.
//
// Widget AppWidgetProvider qui lit les donnees SharedPreferences
// ecrites par WidgetDataService (E5.19a Dart) et affiche
// la progression du trek en cours sur le Home Screen Android.
//
// Donnees lues depuis SharedPreferences (FlutterSharedPreferences) :
// - flutter.widget_trek_trail_name
// - flutter.widget_trek_stage_name
// - flutter.widget_trek_stage_progress (0.0-1.0)
// - flutter.widget_trek_distance_remaining (metres)
// - flutter.widget_trek_eta_minutes
// - flutter.widget_trek_altitude (metres)
// - flutter.widget_trek_stage_index
// - flutter.widget_trek_total_stages
//
// Layout : res/layout/widget_trek_progress.xml
// Config : res/xml/widget_trek_info.xml
// Declare dans AndroidManifest.xml (receiver APPWIDGET_UPDATE)

package com.only1cent.moteur_gr

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.SharedPreferences
import android.widget.RemoteViews

/// Widget Android Home Screen pour la progression trek.
///
/// Lit les donnees de progression depuis SharedPreferences
/// (ecrites par le code Dart via WidgetDataService) et
/// met a jour l'affichage du widget sur le Home Screen.
class TrekWidget : AppWidgetProvider() {

    companion object {
        // Prefixe Flutter SharedPreferences
        private const val PREFS_NAME = "FlutterSharedPreferences"
        private const val PREFIX = "flutter.widget_trek_"
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId, prefs)
        }
    }

    private fun updateAppWidget(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int,
        prefs: SharedPreferences
    ) {
        // Lire les donnees trek depuis SharedPreferences
        val trailName = prefs.getString("${PREFIX}trail_name", "Trek") ?: "Trek"
        val stageName = prefs.getString("${PREFIX}stage_name", "--") ?: "--"
        val stageProgress = prefs.getFloat("${PREFIX}stage_progress", 0f)
        val distanceRemaining = prefs.getFloat("${PREFIX}distance_remaining", 0f)
        val etaMinutes = prefs.getInt("${PREFIX}eta_minutes", 0)
        val altitude = prefs.getFloat("${PREFIX}altitude", 0f)
        val stageIndex = prefs.getInt("${PREFIX}stage_index", 0)
        val totalStages = prefs.getInt("${PREFIX}total_stages", 0)

        // Rendu RemoteViews depuis le layout widget_trek_progress
        val views = RemoteViews(context.packageName, R.layout.widget_trek_progress)
        views.setTextViewText(R.id.trail_name, trailName)
        val stageLabel = if (totalStages > 0) {
            "$stageName ($stageIndex/$totalStages)"
        } else {
            stageName
        }
        views.setTextViewText(R.id.stage_name, stageLabel)
        views.setProgressBar(R.id.progress_bar, 100, (stageProgress * 100).toInt(), false)
        views.setTextViewText(R.id.distance_remaining, "${(distanceRemaining / 1000).toInt()} km")
        views.setTextViewText(R.id.eta, "$etaMinutes min")
        views.setTextViewText(R.id.altitude, "${altitude.toInt()} m")
        appWidgetManager.updateAppWidget(appWidgetId, views)
    }

    override fun onEnabled(context: Context) {
        // Premier widget ajoute au Home Screen
    }

    override fun onDisabled(context: Context) {
        // Dernier widget retire du Home Screen
    }
}

File: data/ref-gr20/R1.10/ios/TrekWidget/TrekWidget.swift
// E5.19b — Widget Home Screen iOS (WidgetKit) — progression trek.
//
// Widget WidgetKit qui lit les donnees SharedPreferences (UserDefaults)
// ecrites par WidgetDataService (E5.19a Dart) et affiche
// la progression du trek en cours sur le Home Screen iOS.
//
// Donnees lues depuis UserDefaults (App Group shared container) :
// - widget_trek_trail_name
// - widget_trek_stage_name
// - widget_trek_stage_progress (0.0-1.0)
// - widget_trek_distance_remaining (metres)
// - widget_trek_eta_minutes
// - widget_trek_altitude (metres)
// - widget_trek_stage_index
// - widget_trek_total_stages
//
// Prerequis : App Group configure dans Xcode (ex: group.com.only1cent.g20App)
// Le widget sera ajoute au projet Xcode comme extension WidgetKit.

import WidgetKit
import SwiftUI

/// Donnees de progression trek pour le widget.
struct TrekProgressEntry: TimelineEntry {
    let date: Date
    let trailName: String
    let stageName: String
    let stageProgress: Double
    let distanceRemaining: Double
    let etaMinutes: Int
    let altitude: Double
    let stageIndex: Int
    let totalStages: Int

    /// Entree placeholder (widget pas encore configure).
    static var placeholder: TrekProgressEntry {
        TrekProgressEntry(
            date: Date(),
            trailName: "GR20",
            stageName: "Etape 3 - Haut Asco",
            stageProgress: 0.45,
            distanceRemaining: 5200,
            etaMinutes: 120,
            altitude: 1422,
            stageIndex: 3,
            totalStages: 16
        )
    }
}

/// Provider de timeline pour le widget trek.
///
/// Lit les donnees depuis UserDefaults (App Group) et
/// fournit les entrees au widget WidgetKit.
struct TrekProgressProvider: TimelineProvider {
    // App Group ID pour partage de donnees Flutter <-> Widget
    private let appGroupId = "group.com.only1cent.g20App"
    private let prefix = "flutter.widget_trek_"

    func placeholder(in context: Context) -> TrekProgressEntry {
        .placeholder
    }

    func getSnapshot(in context: Context, completion: @escaping (TrekProgressEntry) -> Void) {
        completion(readEntry())
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<TrekProgressEntry>) -> Void) {
        let entry = readEntry()
        // Rafraichir toutes les 5 minutes
        let nextUpdate = Calendar.current.date(byAdding: .minute, value: 5, to: Date())!
        let timeline = Timeline(entries: [entry], policy: .after(nextUpdate))
        completion(timeline)
    }

    /// Lit les donnees de progression depuis UserDefaults (App Group).
    private func readEntry() -> TrekProgressEntry {
        guard let defaults = UserDefaults(suiteName: appGroupId) else {
            return .placeholder
        }

        return TrekProgressEntry(
            date: Date(),
            trailName: defaults.string(forKey: "\(prefix)trail_name") ?? "GR20",
            stageName: defaults.string(forKey: "\(prefix)stage_name") ?? "--",
            stageProgress: defaults.double(forKey: "\(prefix)stage_progress"),
            distanceRemaining: defaults.double(forKey: "\(prefix)distance_remaining"),
            etaMinutes: defaults.integer(forKey: "\(prefix)eta_minutes"),
            altitude: defaults.double(forKey: "\(prefix)altitude"),
            stageIndex: defaults.integer(forKey: "\(prefix)stage_index"),
            totalStages: defaults.integer(forKey: "\(prefix)total_stages")
        )
    }
}

/// Vue SwiftUI du widget trek progression.
///
/// Affiche le nom du trail, l'etape en cours, la progression,
/// la distance restante, l'ETA et l'altitude.
struct TrekWidgetEntryView: View {
    var entry: TrekProgressEntry

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            // Titre trek
            Text(entry.trailName)
                .font(.headline)
                .foregroundColor(.white)

            // Etape en cours
            Text("\(entry.stageName) (\(entry.stageIndex)/\(entry.totalStages))")
                .font(.caption)
                .foregroundColor(.white.opacity(0.8))

            // Barre de progression
            ProgressView(value: entry.stageProgress)
                .tint(.green)

            // Stats
            HStack {
                Text("\(Int(entry.distanceRemaining / 1000)) km")
                    .font(.caption2)
                Spacer()
                Text("\(entry.etaMinutes) min")
                    .font(.caption2)
                Spacer()
                Text("\(Int(entry.altitude)) m")
                    .font(.caption2)
            }
            .foregroundColor(.white.opacity(0.7))
        }
        .padding()
        .containerBackground(for: .widget) {
            Color(red: 0.176, green: 0.314, blue: 0.086) // #2D5016
        }
    }
}

/// Definition du widget WidgetKit.
struct TrekWidget: Widget {
    let kind: String = "TrekWidget"

    var body: some WidgetConfiguration {
        StaticConfiguration(kind: kind, provider: TrekProgressProvider()) { entry in
            TrekWidgetEntryView(entry: entry)
        }
        .configurationDisplayName("Progression GR20")
        .description("Suivez votre progression sur le trek en temps reel.")
        .supportedFamilies([.systemSmall, .systemMedium])
    }
}
